packages/eslint-plugin/utils.js

```diff
@@ -116,72 +120,66 @@ function isImported( node ) {
 }
 
 const isTypeFunction = ( type ) => {
-	const functionTypes = [
-		'ArrowFunctionExpression',
-		'FunctionDeclaration',
-		'FunctionExpression',
-	];
-	return functionTypes.includes( type );
+	return FUNCTION_TYPES.includes( type );
 };
 
-function isFunction( node ) {
-	if ( ! node ) {
-		return false;
-	}
+function isFunctionDeclaration( node ) {
+	return node?.type && FUNCTION_TYPES.includes( node.type );
+}
 
-	if ( node?.type === 'Identifier' ) {
-		if (
-			isTypeFunction( node?.parent?.type ) ||
-			isTypeFunction( node?.parent?.init?.type )
-		) {
-			return true;
-		}
-	}
+function isFunctionProperty( node ) {
+	return node.type === 'Property' && node.value && isFunction( node.value );
+}
 
-	const isFunctionDeclaration =
-		node.type &&
-		[
-			'ArrowFunctionExpression',
-			'FunctionDeclaration',
-			'FunctionExpression',
-		].includes( node.type );
+function hasFunctionInDeclarations( declarations ) {
+	return declarations?.some( ( declaration ) =>
+		isFunction( declaration.init )
+	);
+}
 
-	if ( isFunctionDeclaration ) {
-		return true;
+function isFunctionIdentifier( node ) {
+	if ( node?.type !== 'Identifier' ) {
+		return false;
 	}
 
-	if ( node.type === 'Property' && node.value ) {
-		return isFunction( node.value );
-	}
+	return (
+		isTypeFunction( node?.parent?.type ) ||
+		isTypeFunction( node?.parent?.init?.type )
+	);
+}
 
+function isFunctionExportOrVariable( node ) {
 	if (
-		( node.type === 'ExportNamedDeclaration' ||
-			node.type === 'VariableDeclaration' ) &&
-		node.declarations &&
-		node.declarations.length
+		! [ 'ExportNamedDeclaration', 'VariableDeclaration' ].includes(
+			node.type
+		)
 	) {
-		const hasFunctionDeclaration = node.declarations.some(
-			( declaration ) => {
-				return isFunction( declaration.init );
-			}
-		);
-
-		if ( hasFunctionDeclaration ) {
-			return true;
-		}
+		return false;
 	}
 
-	if (
-		( node.type === 'ExportNamedDeclaration' ||
-			node.type === 'VariableDeclaration' ) &&
-		node.declaration
-	) {
+	if ( node.declarations ) {
+		return hasFunctionInDeclarations( node.declarations );
```

Comment:
This needs a tweak to retain the existing logic, as it should only return at this point if the condition is true; when false, it should continue to the next check.

```suggestion
	if ( hasFunctionInDeclarations( node.declarations ) ) {
		return true;
```